File: functions/src/utils/firestore_paths.ts
/**
 * Firestore collection and document path helpers.
 *
 * Mirrors the Dart-side path helpers to keep paths consistent
 * between client and server. Always use these helpers instead of
 * inline string concatenation to avoid path typos.
 *
 * Collection hierarchy:
 *
 *   users/{userId}
 *     └─ notifications/{notificationId}
 *     └─ drafts/{draftId}
 *
 *   groups/{groupId}
 *     ├─ members/{userId}
 *     ├─ expenses/{expenseId}
 *     │   ├─ splits/{splitId}
 *     │   ├─ payers/{payerId}
 *     │   ├─ items/{itemId}
 *     │   └─ attachments/{attachmentId}
 *     ├─ settlements/{settlementId}
 *     ├─ balances/{balancePairId}
 *     └─ activity/{activityId}
 *
 *   friends/{friendPairId}
 *     ├─ expenses/{expenseId}
 *     │   ├─ splits/{splitId}
 *     │   ├─ payers/{payerId}
 *     │   └─ items/{itemId}
 *     ├─ settlements/{settlementId}
 *     ├─ balance/{balanceDocId}
 *     └─ activity/{activityId}
 *
 *   invites/{inviteCode}
 *   userGroups/{userId}/groups/{groupId}
 *   userFriends/{userId}/friends/{friendUserId}
 *   rateLimits/{docId}
 */

// ── Users ─────────────────────────────────────────────────────────────────

export const usersCol = () => "users";
export const userDoc = (userId: string) => `users/${userId}`;

export const userNotificationsCol = (userId: string) =>
  `users/${userId}/notifications`;
export const userNotificationDoc = (userId: string, notificationId: string) =>
  `users/${userId}/notifications/${notificationId}`;

export const userDraftsCol = (userId: string) => `users/${userId}/drafts`;
export const userDraftDoc = (userId: string, draftId: string) =>
  `users/${userId}/drafts/${draftId}`;

// ── Groups ────────────────────────────────────────────────────────────────

export const groupsCol = () => "groups";
export const groupDoc = (groupId: string) => `groups/${groupId}`;

// Members
export const groupMembersCol = (groupId: string) =>
  `groups/${groupId}/members`;
export const groupMemberDoc = (groupId: string, userId: string) =>
  `groups/${groupId}/members/${userId}`;

// Expenses
export const groupExpensesCol = (groupId: string) =>
  `groups/${groupId}/expenses`;
export const groupExpenseDoc = (groupId: string, expenseId: string) =>
  `groups/${groupId}/expenses/${expenseId}`;

// Expense subcollections
export const expenseSplitsCol = (groupId: string, expenseId: string) =>
  `groups/${groupId}/expenses/${expenseId}/splits`;
export const expenseSplitDoc = (
  groupId: string,
  expenseId: string,
  splitId: string
) => `groups/${groupId}/expenses/${expenseId}/splits/${splitId}`;

export const expensePayersCol = (groupId: string, expenseId: string) =>
  `groups/${groupId}/expenses/${expenseId}/payers`;
export const expensePayerDoc = (
  groupId: string,
  expenseId: string,
  payerId: string
) => `groups/${groupId}/expenses/${expenseId}/payers/${payerId}`;

export const expenseItemsCol = (groupId: string, expenseId: string) =>
  `groups/${groupId}/expenses/${expenseId}/items`;
export const expenseItemDoc = (
  groupId: string,
  expenseId: string,
  itemId: string
) => `groups/${groupId}/expenses/${expenseId}/items/${itemId}`;

export const expenseAttachmentsCol = (groupId: string, expenseId: string) =>
  `groups/${groupId}/expenses/${expenseId}/attachments`;
export const expenseAttachmentDoc = (
  groupId: string,
  expenseId: string,
  attachmentId: string
) => `groups/${groupId}/expenses/${expenseId}/attachments/${attachmentId}`;

// Settlements
export const groupSettlementsCol = (groupId: string) =>
  `groups/${groupId}/settlements`;
export const groupSettlementDoc = (
  groupId: string,
  settlementId: string
) => `groups/${groupId}/settlements/${settlementId}`;

// Balances (Cloud Functions only — clients read-only)
export const groupBalancesCol = (groupId: string) =>
  `groups/${groupId}/balances`;
export const groupBalanceDoc = (groupId: string, balancePairId: string) =>
  `groups/${groupId}/balances/${balancePairId}`;

// Activity (Cloud Functions only — clients read-only)
export const groupActivityCol = (groupId: string) =>
  `groups/${groupId}/activity`;
export const groupActivityDoc = (groupId: string, activityId: string) =>
  `groups/${groupId}/activity/${activityId}`;

// ── Friends ───────────────────────────────────────────────────────────────

export const friendsCol = () => "friends";
export const friendDoc = (friendPairId: string) =>
  `friends/${friendPairId}`;

// Friend expenses
export const friendExpensesCol = (friendPairId: string) =>
  `friends/${friendPairId}/expenses`;
export const friendExpenseDoc = (friendPairId: string, expenseId: string) =>
  `friends/${friendPairId}/expenses/${expenseId}`;

// Friend expense subcollections
export const friendExpenseSplitsCol = (
  friendPairId: string,
  expenseId: string
) => `friends/${friendPairId}/expenses/${expenseId}/splits`;
export const friendExpenseSplitDoc = (
  friendPairId: string,
  expenseId: string,
  splitId: string
) => `friends/${friendPairId}/expenses/${expenseId}/splits/${splitId}`;

export const friendExpensePayersCol = (
  friendPairId: string,
  expenseId: string
) => `friends/${friendPairId}/expenses/${expenseId}/payers`;
export const friendExpensePayerDoc = (
  friendPairId: string,
  expenseId: string,
  payerId: string
) => `friends/${friendPairId}/expenses/${expenseId}/payers/${payerId}`;

export const friendExpenseItemsCol = (
  friendPairId: string,
  expenseId: string
) => `friends/${friendPairId}/expenses/${expenseId}/items`;
export const friendExpenseItemDoc = (
  friendPairId: string,
  expenseId: string,
  itemId: string
) => `friends/${friendPairId}/expenses/${expenseId}/items/${itemId}`;

// Friend settlements
export const friendSettlementsCol = (friendPairId: string) =>
  `friends/${friendPairId}/settlements`;
export const friendSettlementDoc = (
  friendPairId: string,
  settlementId: string
) => `friends/${friendPairId}/settlements/${settlementId}`;

// Friend balance (single doc — Cloud Functions only)
export const friendBalanceCol = (friendPairId: string) =>
  `friends/${friendPairId}/balance`;
export const friendBalanceDoc = (
  friendPairId: string,
  balanceDocId: string
) => `friends/${friendPairId}/balance/${balanceDocId}`;

/** Convenience: the single net balance doc for a friend pair. */
export const friendNetBalanceDoc = (friendPairId: string) =>
  `friends/${friendPairId}/balance/net`;

// Friend activity (Cloud Functions only)
export const friendActivityCol = (friendPairId: string) =>
  `friends/${friendPairId}/activity`;
export const friendActivityDoc = (
  friendPairId: string,
  activityId: string
) => `friends/${friendPairId}/activity/${activityId}`;

// ── Invites ───────────────────────────────────────────────────────────────

export const invitesCol = () => "invites";
export const inviteDoc = (inviteCode: string) => `invites/${inviteCode}`;

// ── User Groups (denormalized) ──────────────────────────────────────────

export const userGroupsCol = (userId: string) =>
  `userGroups/${userId}/groups`;
export const userGroupDoc = (userId: string, groupId: string) =>
  `userGroups/${userId}/groups/${groupId}`;

// ── User Friends (denormalized) ─────────────────────────────────────────

export const userFriendsCol = (userId: string) =>
  `userFriends/${userId}/friends`;
export const userFriendDoc = (userId: string, friendUserId: string) =>
  `userFriends/${userId}/friends/${friendUserId}`;

// ── Rate Limits (Cloud Functions only) ──────────────────────────────────

export const rateLimitsCol = () => "rateLimits";
export const rateLimitDoc = (docId: string) => `rateLimits/${docId}`;

File: lib/core/constants/firestore_paths.dart
/// Centralized Firestore collection and document path definitions.
///
/// All Firestore path strings used by data sources should be referenced from
/// this class to avoid typos and ensure consistency. This class cannot be
/// instantiated or extended.
///
/// The Firestore hierarchy is:
/// ```text
/// users/{userId}
///   ├── notifications/{notificationId}
///   └── drafts/{draftId}
/// groups/{groupId}
///   ├── members/{userId}
///   ├── expenses/{expenseId}
///   ├── settlements/{settlementId}
///   ├── balances/{balancePairId}
///   └── activity/{activityId}
/// friends/{friendPairId}
///   ├── expenses/{expenseId}
///   ├── settlements/{settlementId}
///   ├── balance (single doc)
///   └── activity/{activityId}
/// invites/{inviteCode}
/// userGroups/{userId}
/// userFriends/{userId}
/// ```
abstract final class FirestorePaths {
  // ── Top-level collections ────────────────────────────────────────────

  /// Collection of user profile documents.
  static const String users = 'users';

  /// Collection of group documents.
  static const String groups = 'groups';

  /// Collection of friend-pair documents.
  static const String friends = 'friends';

  /// Collection of group invite link documents.
  static const String invites = 'invites';

  /// Collection of per-user group membership lookups.
  static const String userGroups = 'userGroups';

  /// Collection of per-user friend list lookups.
  static const String userFriends = 'userFriends';

  // ── Subcollection names ──────────────────────────────────────────────

  /// Subcollection of group members.
  static const String members = 'members';

  /// Subcollection of expenses (under groups or friends).
  static const String expenses = 'expenses';

  /// Subcollection of settlement records (under groups or friends).
  static const String settlements = 'settlements';

  /// Subcollection of pairwise balance documents (under groups or friends).
  static const String balances = 'balances';

  /// Subcollection of activity feed entries (under groups or friends).
  static const String activity = 'activity';

  /// Subcollection of user notifications.
  static const String notifications = 'notifications';

  /// Subcollection of draft expenses saved locally by the user.
  static const String drafts = 'drafts';

  /// Subcollection of expense splits.
  static const String splits = 'splits';

  /// Subcollection of expense payers.
  static const String payers = 'payers';

  /// Subcollection of itemized expense items.
  static const String items = 'items';

  /// Subcollection of expense attachments (receipts, photos).
  static const String attachments = 'attachments';

  // ── User paths ───────────────────────────────────────────────────────

  /// Returns the document path for a user: `users/{userId}`.
  static String userDoc(String userId) => '$users/$userId';

  /// Returns the notifications subcollection path for a user.
  static String userNotifications(String userId) =>
      '$users/$userId/$notifications';

  /// Returns the drafts subcollection path for a user.
  static String userDrafts(String userId) => '$users/$userId/$drafts';

  // ── Group paths ──────────────────────────────────────────────────────

  /// Returns the document path for a group: `groups/{groupId}`.
  static String groupDoc(String groupId) => '$groups/$groupId';

  /// Returns the members subcollection path for a group.
  static String groupMembers(String groupId) => '$groups/$groupId/$members';

  /// Returns the expenses subcollection path for a group.
  static String groupExpenses(String groupId) => '$groups/$groupId/$expenses';

  /// Returns the settlements subcollection path for a group.
  static String groupSettlements(String groupId) =>
      '$groups/$groupId/$settlements';

  /// Returns the balances subcollection path for a group.
  static String groupBalances(String groupId) => '$groups/$groupId/$balances';

  /// Returns the activity subcollection path for a group.
  static String groupActivity(String groupId) => '$groups/$groupId/$activity';

  // ── Friend paths ─────────────────────────────────────────────────────

  /// Returns the document path for a friend pair: `friends/{friendPairId}`.
  static String friendDoc(String friendPairId) => '$friends/$friendPairId';

  /// Returns the expenses subcollection path for a friend pair.
  static String friendExpenses(String friendPairId) =>
      '$friends/$friendPairId/$expenses';

  /// Returns the settlements subcollection path for a friend pair.
  static String friendSettlements(String friendPairId) =>
      '$friends/$friendPairId/$settlements';

  /// Returns the balance subcollection path for a friend pair:
  /// `friends/{friendPairId}/balance`.
  static String friendBalance(String friendPairId) =>
      '$friends/$friendPairId/balance';

  /// Returns the net balance document path for a friend pair:
  /// `friends/{friendPairId}/balance/net`.
  ///
  /// This is the full document path used by Cloud Functions to store
  /// the computed net balance between two friends.
  static String friendBalanceDoc(String friendPairId) =>
      '$friends/$friendPairId/balance/net';

  /// Returns the activity subcollection path for a friend pair.
  static String friendActivity(String friendPairId) =>
      '$friends/$friendPairId/$activity';

  // ── Utilities ────────────────────────────────────────────────────────

  /// Generates a canonical friend pair ID from two user IDs.
  ///
  /// The pair ID is deterministic: `min(userA, userB)_max(userA, userB)`.
  /// This ensures the same pair always maps to the same document regardless
  /// of argument order.
  ///
  /// Example:
  /// ```dart
  /// FirestorePaths.friendPairId('bob', 'alice'); // → 'alice_bob'
  /// FirestorePaths.friendPairId('alice', 'bob'); // → 'alice_bob'
  /// ```
  static String friendPairId(String userA, String userB) {
    final sorted = [userA, userB]..sort();
    return '${sorted[0]}_${sorted[1]}';
  }
}
